README.md
# nodejs-poolController-webClient

This client application is for use with [nodejs-poolController](https://github.com/tagyoureit/nodejs-poolcontroller) (an app for controlling Pentair pool equipment).

## Installation

If you want to use SSDP for automatic configuration then please ensure it is enable on the nodejs-poolController app (in config.json).

To use this client, install it as a separate application:
1. Download or clone the repository
1. `npm i` to install dependencies
1. Optional - modify `config.json`:
    * Change the port where end users will load this client.  Default is `8080`.
1. Start the app. 
    * `npm start` will clean out the compiled directories and recompile all files anew.  This should be the starting script every time you download/pull a new version.  
    * For faster startup times, run `npm run start:cached` to use the existing compiled files created by `npm start`.
1. To manually enter the ip of nodejs-poolController open the hamburger menu (top right) and disable SSDP and enter an address.

nodejs-poolController by default will only listen to connections on the localhost (127.0.0.1) adapter.  If you will be running this client on a different machine edit `config.json` in the nodejs-poolController app so http/https servers listen on a different (external) interface.  

# Virtual Controllers
* [Virtual Chlorinator Controller Directions](https://github.com/tagyoureit/nodejs-poolController/wiki/Virtual-Chlorinator-Controller-v6)
* [Virtual Pump Controller Directions](https://github.com/tagyoureit/nodejs-poolController/wiki/Virtual-Pump-Controller---v6)

Changelog

# 6.1.0
1. Bumped version by 5.1 to match core platform
1. Connection to njsPC is through it's own server for this app.  Prior version would retrieve the ip of the njsPC from the webClient server and then the browser would connect directly to njsPC.  This wasn't ideal for numerous reasons, including proxies and remote connections, so now the app connects through the server included here.
1. Chem Controllers support
1. UI for "server" time which keeps OCP (outdoor control panel) sync with host time in cases (*Touch) where it doesn't make it's own internet connection
1. Version notification for core platform
1. Many bug fixes
1. Personal/location information


# 1.0.0
Initial Release - compatible with nodejs-poolController 6.0.0

src/server.ts
import * as path from 'path';
import Bundler=require('parcel-bundler');
import express=require('express');
import { config } from './Config';

var ssdp=require('node-ssdp').Client
    , client=new ssdp({});
const mSearch='urn:schemas-upnp-org:device:PoolController:1';
let configURL: string=config.getSection("discovery").poolControllerURL;
let discoveredURL: string=configURL;
let timeout: NodeJS.Timeout;

function search(force?: boolean) {
    if(configURL==="*"||force) {
        if(typeof timeout!==undefined) clearTimeout(timeout);
        client.search(mSearch);
        timeout=setTimeout(function() {
            console.log('restarting search');
            client.stop();
            search();
        }, 5000);
    }
}

async function startBundler() {
    const app=express();
    client.on('response', function inResponse(headers, code, rinfo) {
        if(headers.ST!==mSearch||configURL!=='*') { return; }
        clearTimeout(timeout);
        console.log('Got a response to an m-search:\n%d\n%s\n%s', code, JSON.stringify(headers, null, '  '), JSON.stringify(rinfo, null, '  '));
        // console.log(headers.LOCATION);
        const oldUrl=discoveredURL;
        discoveredURL=headers.LOCATION.match(/^.+?[^\/:](?=[?\/]|$)/)[0];
        if(discoveredURL!=='*'&&discoveredURL!==oldUrl) console.log(`Found pool server at new address ${ discoveredURL } (previously was ${ oldUrl })`)
        client.stop();
    });

    app.use(function(req, res, next) {
        res.header("Access-Control-Allow-Origin", "*"); // update to match the domain you will make the request from
        res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
        res.header('Access-Control-Allow-Methods', 'OPTIONS, GET, POST, PUT, DELETE');
        next();
    });
    app.use(express.json());
    app.get('/discover', (req, res) => {
        if(configURL!=='*') {
            res.status(200).send({ url: configURL });
        }
        else if(discoveredURL==="*") {
            console.log(`SSDP: cannot find poolController with discovery.  Please set address manually in config.json in the format of 'http://ip:port/`);
            res.status(204).send(`No SSDP found yet.`);
            client.search(mSearch);
        }
        else {
            console.log(`about to send: ${ discoveredURL }`);
            res.status(200).send({ url: discoveredURL });
        }
    });

    app.get('/recheck', (req, res) => {
        if(configURL!=='*') {
            console.log(`Webapp not able to connect to ${ configURL }.`)
        }
        else {
            console.log(`Webapp not able to connect; forcing mSearch.`)
            client.search(mSearch, true);
        }
    })

    app.get('/visibility', (req, res) => {
        res.send(config.getSection('options').hidePanels)
    })
    app.delete('/panel', (req, res) => {
        let options=config.getSection("options")
        options.hidePanels = [];
        res.send(config.setSection('options', options))
    })
    app.put('/panel', (req, res) => {
        let options=config.getSection("options")
        let { hidePanels }=options;
        const { name, state }=req.body;
        if(state==='hide') {

            if(!hidePanels.includes(name)) {
                hidePanels.push(name);
            }
        }
        else {
            const index=hidePanels.indexOf(name);
            if(index!==-1) {
                hidePanels.splice(index, 1);
            }
        }
        console.log(JSON.stringify(options, null, 2))
        config.setSection("options", options);
    })

    // Parcel: absolute path to entry point
    const file=path.join(process.cwd(), './web/dashboard/index.html');
    console.log(`Parcel serving files from: ${ file }`);
    // Parcel: set options
    const options={
        outDir: './dist/web'
    };
    // Parcel: Initialize a new bundler
    const bundler=new Bundler(file, options);
    // Let express use the bundler middleware, this will let Parcel handle every request over your express server
    app.use(bundler.middleware());

    // Listen on port 8080
    app.listen(config.getSection("web").port);
    search();
}

startBundler();

web/dashboard/app.js
"use strict";
exports.__esModule = true;
var React = require("react");
var ReactDOM = require("react-dom");
var react_router_dom_1 = require("react-router-dom");
var PoolController_1 = require("./components/PoolController");
var UtilitiesLayout_1 = require("./components/utilities/UtilitiesLayout");
var PacketSnifferController_1 = require("./components/utilities/PacketSnifferController");
var PacketTester_1 = require("./components/utilities/PacketTester");
var Replay_1 = require("./components/utilities/Replay");
var App = function () {
    return (<PoolController_1["default"] />);
};
var Utilities = function () {
    return (<UtilitiesLayout_1["default"] />);
};
var packetSniffer = function () {
    return (<PacketSnifferController_1["default"] />);
};
var packetTester = function () {
    return (<PacketTester_1["default"] />);
};
var replay = function () {
    return (<Replay_1["default"] />);
};
ReactDOM.render(<react_router_dom_1.BrowserRouter>
        <react_router_dom_1.Route exact path="/" component={App}/>
        <react_router_dom_1.Route path="/packetSniffer" component={packetSniffer}/>
        <react_router_dom_1.Route path="/utilities" component={Utilities}/>
        <react_router_dom_1.Route path='/packetTester' component={packetTester}/>
        <react_router_dom_1.Route path='/replay' component={replay}/>
    </react_router_dom_1.BrowserRouter>, document.getElementById('root'));


scripts/smartthings.js
var http = require('http');
var io = require('socket.io-client');
var patch = require('socketio-wildcard')(io.Manager);
// var configFile = container.settings.getConfig()

var address = '127.0.0.1';
var port = 8080;
// var address = '10.0.0.39';
// var port = 39501;



var serverURL;

// if (configFile.outputToSmartThings.hasOwnProperty("logEnabled")) {
//     logEnabled = configFile.outputToSmartThings.logEnabled
// }


// if (secureTransport) {
//     serverURL = 'https://localhost:' + configFile.poolController.https.expressPort
// } else {
//     serverURL = 'http://localhost:' + configFile.poolController.http.expressPort
// }
serverURL = 'http://localhost:4200'


var socket = io.connect(serverURL, {
    secure: false,
    // secure: secureTransport,
    reconnect: true,
    rejectUnauthorized: false
});
patch(socket);
function notify(event, data) {
    if (address !== '*') {
        var json = JSON.stringify(data);

        var opts = {
            method: 'NOTIFY',
            host: address,
            port: port,
            path: '/notify',
            headers: {
                'CONTENT-TYPE': 'application/json',
                'CONTENT-LENGTH': Buffer.byteLength(json),
                'X-EVENT-TYPE': event
            }
        };

        var req = http.request(opts);
        req.on('error', function (err, req, res) {
           console.error(err);
        });
        req.write(json);
        req.end();
        // if (logEnabled) {
            //if (event === "tempBody" || event === "body")
            //console.debug('outputToSmartThings sent event %s', event);
            console.log('outputToSmartThings (' + address + ':' + port + ') Sent ' + event + "'" + json + "'");
        // }
    }

}

socket.on('*', function (data) {
    notify(data.data[0], data.data[1]);
});





src/Server.ts
require('source-map-support').install();
import axios, { AxiosRequestConfig, Method } from 'axios';
import * as http from "http";
import * as path from 'path';
import * as io from 'socket.io-client';
import { config } from './Config';
import Bundler = require('parcel-bundler');
import express = require('express');

import ioServer = require("socket.io");

var ssdp = require('node-ssdp').Client
    , client = new ssdp({});
const mSearch = 'urn:schemas-upnp-org:device:PoolController:1';
let { override = { protocol: 'http://', host: 'localhost', port: 4200 }, useSSDP = true, poolURL = '' }: { override: { protocol: string, host: string, port: number }, useSSDP: boolean, poolURL?: string } = config.getSection("discovery");
let { protocol, host, port }: { protocol: string, host: string, port: number } = override;
let timeout: NodeJS.Timeout;

let ioClient: SocketIOClient.Socket, patch;
let server: http.Server;
let sockServer: ioServer.Server;


function search(force?: boolean) {
    if (useSSDP || force) {
        if (typeof timeout !== undefined) clearTimeout(timeout);
        client.search(mSearch);
        timeout = setTimeout(function () {
            console.log('restarting search');
            client.stop();
            search();
        }, 5000);
    }
}

const reloadVars = () => {
    let d = config.getSection('discovery');
    override = d.override;
    useSSDP = d.useSSDP;
    poolURL = d.poolURL;
    protocol = d.override.protocol;
    host = d.override.host;
    port = d.override.port;
    if (!useSSDP) {
        if (typeof poolURL === 'undefined' || poolURL !== `${protocol}://${host}:${port}`) {
            let { protocol, host, port } = override;
            poolURL = `${protocol}://${host}:${port}`;
            saveVars();
        }
    }
    initSocketClient();
}
const saveVars = () => {
    let d = config.getSection('discovery');
    d.override = override;
    d.useSSDP = useSSDP;
    d.poolURL = poolURL;
    d.override.protocol = protocol;
    d.override.host = host;
    d.override.port = port;
    config.setSection('discovery', d);
    initSocketClient();
}

const initSocketClient = () => {
    if (typeof ioClient !== 'undefined') ioClient.disconnect();
    ioClient = io.connect(poolURL);
    patch = require('socketio-wildcard')(io.Manager);
    patch(ioClient);
    ioClient.on('connect_error', (data) => {
        console.log(`SOCKET connect error.`)
    });
    ioClient.on('disconnect', (data) => {
        console.log(`SOCKET disconnect.`)
    });
    ioClient.on('reconnect', (data) => {
        console.log(`SOCKET reconnect.`)
    });
    ioClient.on('*', (socketData) => {
        // console.log(`Incoming socket: ${socketData.data[0]} with data`)
        // console.log(socketData.data[1])
        if (socketData.data[1] === null || socketData.data[1] === undefined) {
            console.log(`ALERT: Null socket data received for ${socketData.data[0]}`);
        } else {
            // emitter.emit(data.data[0], data.data[1]);
            sockServer.emit(socketData.data[0], socketData.data[1]);
        }
    });
}
const initSocketServer = () => {
    var middleware = require('socketio-wildcard')();
    sockServer = ioServer(server, { cookie: true });
    sockServer.use(middleware);
    sockServer.on('error', (err) => {
        console.error('Socket server error %s', err.message);
    });
    sockServer.on('connect_error', (err) => {
        console.error('Socket connection error %s', err.message);
    });
    sockServer.on('reconnect_failed', (err) => {
        console.error('Failed to reconnect with socket %s', err.message);
    });
    sockServer.on('connection', (sock: ioServer.Socket) => {
        console.log(`New socket client connected ${sock.id} -- ${sock.client.conn.remoteAddress}`);
        sock.on('*', (socketData) => {
            console.log(`Passthrough socket: ${socketData.data[0]} with data`)
            console.log(socketData.data[1])
            ioClient.emit(socketData.data[0], socketData.data[1]);
        })
    });
}

async function startBundler() {
    const app = express();
    server = http.createServer(app);

    client.on('response', function inResponse(headers, code, rinfo) {
        if (headers.ST !== mSearch || !useSSDP) { return; }
        clearTimeout(timeout);
        console.log('Got a response to an m-search:\n%d\n%s\n%s', code, JSON.stringify(headers, null, '  '), JSON.stringify(rinfo, null, '  '));
        // console.log(headers.LOCATION);
        const oldUrl = poolURL;
        poolURL = headers.LOCATION.match(/^.+?[^\/:](?=[?\/]|$)/)[0];
        console.log(`IN SEARCH --- ${poolURL}`)
        if (typeof poolURL !== 'undefined' && poolURL !== oldUrl) console.log(`Found pool server at new address ${poolURL} (previously was ${oldUrl})`)

        saveVars();
        client.stop();
    });

    app.use(function (req, res, next) {
        res.header("Access-Control-Allow-Origin", "*"); // update to match the domain you will make the request from
        res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
        res.header('Access-Control-Allow-Methods', 'OPTIONS, GET, POST, PUT, DELETE');
        console.log(`${req.method} ${req.path}`)
        next();
    });
    app.use(express.json());

    app.get('/discover', (req, res) => {
        reloadVars();
        if (!useSSDP) {
            // res.status(200).send({ url: `${protocol}://${host}:${port}` });
        }
        else if (useSSDP && typeof poolURL === 'undefined') {
            console.log(`SSDP: cannot find poolController with discovery.  Please set address manually in config.json in the format of 'http://ip:port/`);
            client.search(mSearch);
        }
        res.status(200).json({ override, useSSDP, poolURL });
    });

    app.get('/recheck', (req, res) => {
        reloadVars();
        if (!useSSDP) {
            console.log(`Config.json set to override SSDP: ${poolURL}.`)
        }
        else {
            console.log(`Webapp not able to connect; forcing mSearch.`)
            client.search(mSearch, true);
        }
        return res.status(200);
    })

    app.get('/visibility', (req, res) => {
        res.send(config.getSection('options').hidePanels)
    })
    app.delete('/panel', (req, res) => {
        let options = config.getSection("options")
        options.hidePanels = [];
        res.send(config.setSection('options', options))
    })
    app.put('/panel', (req, res) => {
        let options = config.getSection("options")
        let { hidePanels } = options;
        const { name, state } = req.body;
        if (state === 'hide') {

            if (!hidePanels.includes(name)) {
                hidePanels.push(name);
            }
        }
        else {
            const index = hidePanels.indexOf(name);
            if (index !== -1) {
                hidePanels.splice(index, 1);
            }
        }
        console.log(JSON.stringify(options, null, 2))
        config.setSection("options", options);
        reloadVars();
        res.send(config.setSection('options', options))
    })
    app.get('/chemController', (req, res) => {
        res.send(config.getSection('chemController'));
    })
    app.put('/chemController', (req, res) => {
        // let cc=config.getSection("chemController") || {"chemController": {}}
        // if (typeof req.body.alarms !== 'undefined') cc.alarms = req.body.alarms;
        // if (typeof req.body.warnings !== 'undefined') cc.warnings = req.body.warnings;
        config.setSection("chemController", req.body);
        res.send(req.body);
    })
    app.put('/discovery', (req, res) => {
        protocol = req.body.override.protocol;
        host = req.body.override.host;
        port = parseInt(req.body.override.port, 10);
        useSSDP = req.body.useSSDP;
        if (useSSDP) {
            poolURL = undefined;
            saveVars();
            search();
        }
        else {
            poolURL = `${protocol}://${host}:${port}`;
            saveVars();
            client.stop();
        }
        res.status(200).json({ override, useSSDP, poolURL });
    })


    app.all('*', async (req, res, next) => {
        let paths: string[];
        if (req.url.startsWith('/')) paths = req.url.substring(1).split('/');
        else paths = req.url.split('/');
        switch (paths[0]) {
            case 'state':
            case 'config':
            case 'extended':
            case 'app':
                {
                    console.log(`Redirecting request to : ${poolURL}${req.path}`)
                    let opts: AxiosRequestConfig = {
                        // url: `${req.protocol}://${req.hostname}:4200${req.path}`,
                        url: `${poolURL}${req.path}`,
                        method: req.method as Method,
                        data: req.body
                    }
                    try {
                        let axiosRes = await axios(opts);
                        res.status(200).send(axiosRes.data);
                    }
                    catch (err) {
                        console.log(`Error fetching data: ${err.message}`);
                        console.log(err);
                        res.status(500).send(err.message);
                    }
                    break;
                }
            default:
                next();
        }
    })

    initSocketClient();
    initSocketServer();
    if (process.env.NODE_ENV !== 'production') {
        // Parcel: absolute path to entry point
        const file = path.join(process.cwd(), './web/dashboard/index.html');
        console.log(`Parcel serving files from: ${file}`);
        // Parcel: set options
        const options = {
            outDir: './dist/web'
        };
        // Parcel: Initialize a new bundler
        const bundler = new Bundler(file, options);
        // Let express use the bundler middleware, this will let Parcel handle every request over your express server
        app.use(bundler.middleware());
    }
    else {
        app.use(express.static(path.join(__dirname, '/web')));
    }
    // Listen on port 8080
    server.listen(config.getSection("web").port);
    search();
}

startBundler(); 
